src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs

```diff
@@ -580,5 +582,639 @@ public void HangingLambdaParsing_Bug14167()
+                {
+                    N(SyntaxKind.IdentifierToken, "func1");
+                }
+                N(SyntaxKind.EqualsValueClause);
+                {
+                    N(SyntaxKind.EqualsToken);
+                    N(SyntaxKind.ParenthesizedExpression);
+                    {
+                        N(SyntaxKind.OpenParenToken);
+                        N(SyntaxKind.LogicalNotExpression);
+                        {
+                            N(SyntaxKind.ExclamationToken);
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        N(SyntaxKind.CloseParenToken);
+                    }
+                }
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+
+        [Fact]
+        public void TestNullCheckedDiscard()
+        {
+            UsingDeclaration("Func<int> func1 = (_!) => 42;");
+            N(SyntaxKind.FieldDeclaration);
+            {
+                N(SyntaxKind.VariableDeclaration);
+                {
+                    N(SyntaxKind.GenericName);
+                    N(SyntaxKind.IdentifierToken, "Func");
+                    N(SyntaxKind.TypeArgumentList);
+                    {
+                        N(SyntaxKind.LessThanToken);
+                        N(SyntaxKind.PredefinedType);
+                        {
+                            N(SyntaxKind.IntKeyword);
+                        }
+                        N(SyntaxKind.GreaterThanToken);
+                    }
+                }
+                N(SyntaxKind.VariableDeclarator);
+                {
+                    N(SyntaxKind.IdentifierToken, "func1");
+                }
+                N(SyntaxKind.EqualsValueClause);
+                {
+                    N(SyntaxKind.EqualsToken);
+                    N(SyntaxKind.ParenthesizedLambdaExpression);
+                    {
+                        N(SyntaxKind.ParameterList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.Parameter);
+                            {
+                                N(SyntaxKind.IdentifierToken, "_");
+                                N(SyntaxKind.ExclamationToken);
+                            }
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                        N(SyntaxKind.EqualsGreaterThanToken);
+                        N(SyntaxKind.NumericLiteralExpression);
+                        {
+                            N(SyntaxKind.NumericLiteralToken);
+                        }
+                    }
+                }
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+
+        [Fact]
+        public void TestAnonymousDelegateNullChecking()
+        {
+            UsingTree(@"
+delegate void Del(int x!);
```

Comment:
this should likely be an error right?  you shouldnt' be allowed to place these on delegate-decls or interface/abstract methods right?

Reply:
Note: if these are not allowed, i would put the check for this *not* in hte parser.

Reply:
Will create interface/abstract tests and take a look at delegates

Reply:
Note: just to make sure we're on hte same page, even if this is not allowed by the language, the checks should *not* be in the parser.  That will make parsing of these guys even more context sensitive, which adds complexity and makes incremental parsing harder/more-dangerous.  So we should allow parsing of these, but then give a nice error later on (i.e. when making the actual parameter symbols) if these are not allowed for teh current sort of construct.

Another case: `extern` methods.

Reply:
if you are already writing tests for abstract methods and interface methods then please go ahead, but I would approve the PR without them as I know you'll get to them when you do the semantic side of this feature.

In general please feel free to handle suggestions for additions with `// PROTOTYPE` comments, as long as the implementation is correct for the scenarios you've tested so far.